quebecats.py

```diff
@@ -185,39 +190,38 @@ def tick_epilogue(t):
             final_range += 1
             t += 1
 
+        firing_ticks.append(t)
         if display:
-            from sprite_display import draw_grid_frames_and_sprites
-
-            def cell_draw(ctx, i):
-                if not ctx.model.eval(ctx.cell.var.has_mirror):
-                    return
-                color = (0, 0, 1, 1)
-                if ctx.model.eval(ctx.cell.var.mirror_state(ctx.t)):
-                    ctx.line(0.2, 0.8, 0.8, 0.2, color=color)
-                else:
-                    ctx.line(0.2, 0.2, 0.8, 0.8, color=color)
-
-            def edge_draw(ctx, i):
+            def cell_draw(ctx):
+                frame, t_start, t_end = ctx.extra
+                draw_sprite(ctx, laser, t_start, t_end, laser_draw)
```

Comment:
I think this is our fundamental disagreement about sprites. I see them as quite fundamentally tied to cells at each tick, so it's a reasonable thing to say "while drawing a cell, also draw any sprites that happen to pass through this cell". But I do see your point of view. I'll think about this.

Reply:
I have been thinking about adding the grid itself as a first-class "grid part" for other reasons; basically supplying a hook that means "translate/rescale the display context and then just call this once, for the entire grid, before messing around with individual parts". This would be another natural place to draw sprites, if you prefer to think of them being attached to the whole grid rather than an individual part. Would that make more sense to you?

Reply:
Ehhh, maybe? If the context passed to that hook had methods for drawing sprites that look similar to the original sprite context, I could imagine getting behind that possibly. But I think once you've put sprite-use-case-specific methods on the grid-part context, you're already halfway to treating sprites as first-class things themselves anyway, aren't you? I don't see the harm in going all the way then.

Here's an argument for not doing it in cells though: imagine we complete the set of regular planar tilings by adding a `TriangularGrid`. Just like you rotate the point context 180° for half of the points in the hex grid, I'm guessing you'll probably want to rotate the cell context 180° for half of the cells in the triangle grid (although watch out if you're drawing text! Then you're back to un-Cairoing what you've so cleverly Cairoed again.). But you don't want to do that with sprite rendering, or half the time the sprites will look like they're wandering out the opposite vertex to the side they're actually walking out.